c10/core/TensorImpl.h

```diff
@@ -1564,31 +1564,55 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
    * can be validly read from this tensor.
    */
   inline void* mutable_data() {
-    return mutable_data_impl();
+    return data_impl<decltype(*this)>{}(*this, &Storage::mutable_data);
   }
 
  private:
-  // Templated implementation of data() and mutable_data().
+  // Shared implementation of data() and mutable_data() regardless of
+  // constness.
```

Comment:
is the template really worth dealing with the TORCH_CHECK. Couldn't you have just factored the TORCH_CHECKs into a helper method

Reply:
Up to you. But there's still the pointer math duplication.